maps/wasteland/pvp_shield.lua

```diff
@@ -227,70 +227,127 @@ local function on_player_changed_position(event)
+        local enttiy_force = entity.force
+        for _, shield in pairs(this.pvp_shields) do
+            if entity_surface == shield.surface then
+                local box = shield.box
+                local left_top = box.left_top
+                local right_bottom = box.right_bottom
+                if left_top.x <= x and right_bottom.x >= x and left_top.y <= y and right_bottom.y >= y then
+                    local shield_force = shield.force
+                    if enttiy_force == shield_force
+                        and not (shield_force == cause_force or shield_force.get_friend(cause_force) or shield_force.get_cease_fire(cause_force))
+                    then
+                        return true
+                    end
+                end
+            end
+        end
     end
+
+    return false
 end
 
-local shield_disallowed_vehicles = {'tank', 'car'}
+local _unit_filter = {type = "unit", area = nil, force = "enemy"}
 local function scan_protect_shield_area()
     -- Handle edge case damage situations
 
     local this = ScenarioTable.get_table()
+    local all_pvp_vehicles = this.all_pvp_vehicles or {}
     local limit_idx = 0
-    local update_limit = 10
+    local vehicle_update_limit = 10
+    local unit_update_limit = 60
     for _, shield in pairs(this.pvp_shields) do
-        if game.tick % update_limit == limit_idx % update_limit then  -- Keep runtime low
+        if game.tick % vehicle_update_limit == limit_idx % vehicle_update_limit then  -- Keep runtime low
 
             -- Protect against rolling tanks where player hops out before impact - this cannot be handled with damage event
             local tank_box = enlarge_bounding_box(shield.box, 3)
-            for _, e in pairs(shield.surface.find_entities_filtered({name = shield_disallowed_vehicles, area = tank_box })) do
-                if not is_allowed_in_shield(shield, e.force) then
-                    e.speed = 0
+            local tank_box_left_top_x = tank_box.left_top.x
+            local tank_box_left_top_y = tank_box.left_top.y
+            local tank_box_right_bottom_x = tank_box.right_bottom.x
+            local tank_box_right_bottom_y = tank_box.right_bottom.y
+            local i = 0
+            while true do
+                i = i + 1
+                local e = all_pvp_vehicles[i] -- LuaEntity
+                if e then
+                    if e.valid then
+                        local p = e.position
+                        local e_force = e.force
+                        if (p.x > tank_box_left_top_x and p.y > tank_box_left_top_y
+                            and p.x < tank_box_right_bottom_x and p.y < tank_box_right_bottom_y)
+                            and not (shield.force == e_force or shield.force.get_friend(e_force) or shield.force.get_cease_fire(e_force))
+                        then
+                            e.speed = 0
+                        end
+                    else
+                        table.remove(all_pvp_vehicles, i)
```

Comment:
You remove table items while iterating - this is working, yeah? I don't know if the indexes change during iteration.

Also if there's a gap in the indexes, it will stop the whole loop (break later). If one vehicle is removed, can the iteration stop too early?

Reply:
https://github.com/blubFisch/Wasteland/blob/b49faa42ce628a37712fdd867e86b9d44391277f/maps/wasteland/pvp_shield.lua#L311 I do change index

Reply:
> if there's a gap

I don't create gaps.

Reply:
Sure there can't be a gap? Like if you remove index 3 the iteration will stop at the next run and index 4 will not be visited? Or are the gaps closed?

Reply:
This could create a bug that takes days to fix